modules/diskq/logqueue-disk-non-reliable.c

```diff
@@ -223,7 +225,9 @@ _ack_backlog(LogQueue *s, gint num_msg_to_ack)
       if (self->qbacklog->length < ITEM_NUMBER_PER_MESSAGE)
         return;
       msg = g_queue_pop_head(self->qbacklog);
-      POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qbacklog), &path_options);
+      next_msg = g_queue_pop_head(self->qbacklog);
+      gint size = GPOINTER_TO_INT(next_msg);
+      POINTER_TO_LOG_PATH_OPTIONS(size, &path_options);
```

Comment:
This warning type is tricky as it warns on casts, which are usually deliberate decisions.
`POINTER_TO_LOG_PATH_OPTIONS()` contains such an intentional cast, hence the problem.

`Wbad-function-cast` can be useful for detecting changed interfaces at API separation points, but in this special case, it just causes problems.

Removing this macro is not a good idea as it makes the code hard to read.
Maybe a reimplementation could help, for example:
```
#define POINTER_TO_LOG_PATH_OPTIONS(ptr, lpo) ({ \
  gpointer p = ptr;
  LogPathOptions *opts = lpo;
  (lpo)->ack_needed = (GPOINTER_TO_INT(p) & ~0x80000000) \
})
```

Or even better: a real C function could work too.

Reply:
Fixed as desired 👍